tsl/src/remote/connection.c

```diff
@@ -2214,18 +2223,96 @@ send_end_binary_copy_data(const TSConnection *conn, TSConnectionError *err)
 	return true;
 }
 
+/*
+ * End COPY on the remote connection.
+ * This function is used to clean up after errors as well, so it works in a
+ * recovery fashion: it tries to bring the connection into predictable and
+ * usable state, even if there are some errors or discrepancies between its
+ * actual and expected state.
+ */
 bool
 remote_connection_end_copy(TSConnection *conn, TSConnectionError *err)
 {
 	PGresult *res;
 	bool success;
 
+	/*
+	 * In any case, try to switch the connection into the blocking mode, because
+	 * that's what the non-COPY code expects.
+	 */
+	if (PQisnonblocking(conn->pg_conn))
+	{
+		/*
+		 * We have to flush the connection before we can switch it into blocking
+		 * mode.
+		 */
+		for (;;)
+		{
+			CHECK_FOR_INTERRUPTS();
+
+			int flush_result = PQflush(conn->pg_conn);
+
+			if (flush_result == 1)
+			{
+				/*
+				 * The socket is busy, wait. We don't care about the wait result
+				 * here, because whether it is a timeout or the socket became
+				 * writeable, we just retry.
+				 */
+				(void) WaitLatchOrSocket(MyLatch,
+										 PQsocket(conn->pg_conn),
+										 WL_TIMEOUT | WL_SOCKET_WRITEABLE,
+										 /* timeout = */ 1000,
+										 /* wait_event_info = */ 0);
+			}
+			else if (flush_result == 0)
+			{
+				/* Flushed all. */
+				break;
+			}
+			else
+			{
+				/* Error. */
+				return fill_simple_error(err,
+										 ERRCODE_CONNECTION_EXCEPTION,
+										 "failed to flush the COPY connection",
+										 conn);
+			}
+		}
+
+		/* Switch the connection into blocking mode. */
+		if (PQsetnonblocking(conn->pg_conn, 0))
```

Comment:
Again, the success check for this command seems inconsistent with `PQisnonblocking()`. I would recommend explicit return value check.